components/tidb_query_executors/src/runner.rs

```diff
@@ -511,13 +525,16 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
         // record count less than batch size and is not drained
         while record_len < self.stream_row_limit && !is_drained {
             let mut current_chunk = Chunk::default();
-            let (drained, len) = self.internal_handle_request(
+            let (drained, len, quota_delay) = self.internal_handle_request(
                 true,
                 batch_size.min(self.stream_row_limit - record_len),
                 &mut current_chunk,
                 &mut warnings,
                 &mut ctx,
             )?;
+
+            std::thread::sleep(quota_delay);
```

Comment:
Thread sleep is used here but an async sleep is used everywhere else. How do I know which to choose?

Reply:
use async sleep in aysnc code block and thread sleep in sync code block

Reply:
But the effect is different. If an async sleep is used, the same thread pool can still handle other requests immediately. But a thread sleep will truly block the thread, making the thread unable to serve any task for a while. Which one is you really want?

Reply:
seems that we want a truly block

Reply:
I think it's better to use async style sleep. The reason is that it's very expensive to frequently call OS's sleep as it will do context switch. Second is that we may want to let some requests run as the thread may run different requests. Third,  by the test it turns out that async sleep works pretty well. 
The way the async sleep still works is to slow down the requests so that the caller side will slow down. Because the caller won't send unlimited requests to TiKV without waiting current ones back.